chore(Dockerfile): remove chgrp/chown layer

npepinpe opened the pull request:
## Description

This PR slims down the Docker image by removing an unnecessary layer which was > 100Mb. When moving to non-root, we introduced a chgrp/chmod layer which effectively has the same size as the copy layer, resulting two large layers that need to be pushed on every change. Furthermore, it seems the container was still not running under the `zeebe` user by default, which is fixed, and the root folder is also given the correct permissions. This was done by changing the builder stage to a debian based image, and using the same permissions, which are they copied over properly (tested with and without BuildKit).

## Related issues

closes #4788 

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

npepinpe commented:
Assigned to Zell as well as I'm not sure if @menski has time for this.

Zelldon commented on Dockerfile:
```diff
@@ -24,21 +29,22 @@ ENV ZB_HOME=/usr/local/zeebe \
 
+RUN groupadd -g 1000 zeebe && \
+    adduser -u 1000 zeebe --system --ingroup zeebe
+RUN chmod g=u /etc/passwd
```
I'm not a docker expert. I think earlier it was told that RUN should be merged together, but it seems that this is in the newest versions no longer the case, is this right? Related https://stackoverflow.com/questions/39223249/multiple-run-vs-single-chained-run-in-dockerfile-what-is-better

npepinpe replied:
You're right that it's usually better to merge them, in this case since those layers tend to be tiny and not updated often I didn't, but we could stay consistent. The line about it being less important in newer versions refers to the multi stage - since you don't care how many layers the non-target stages have, and having multiple run statements is usually more readable :man_shrugging:

npepinpe commented on Dockerfile:
```diff
@@ -24,21 +29,22 @@ ENV ZB_HOME=/usr/local/zeebe \
+RUN chown 1000:0 ${ZB_HOME} ${ZB_HOME}/data
+
+USER zeebe
 COPY --chown=1000:0 docker/utils/startup.sh /usr/local/bin
```
No, we probably should, good catch :+1:

npepinpe commented on Dockerfile:
```diff
@@ -12,9 +12,14 @@ RUN mkdir -p ${TMP_DIR} && \
 ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini ${TMP_DIR}/bin/tini
 COPY docker/utils/startup.sh ${TMP_DIR}/bin/startup.sh
 RUN chmod +x -R ${TMP_DIR}/bin/
+RUN chgrp 0 ${TMP_DIR}/bin/startup.sh && chmod 0775 ${TMP_DIR}/bin/startup.sh
```
I essentially moved this line, so I didn't think too much about it and trusted whoever added it knew why - but you're right, 0 is root so no idea why we're changing the group to root :thinking: We're anyway later changing it again to 1000, so yeah, not sure. I guess I should test this stuff on OpenShift since this is where the requirement came from

npepinpe commented:
I rebased and updated based on your review, thanks :slightly_smiling_face: Should be a bit simpler now, and the nice thing is the copy of our app is the last layer!

So image size goes from 450Mb to 301Mb :slightly_smiling_face:

menski commented on Dockerfile:
```diff
@@ -16,9 +15,13 @@ RUN mkdir -p ${TMP_DIR} && \
     # already create volume dir to later have correct rights
     mkdir ${TMP_DIR}/data
 
+RUN groupadd -g 1000 zeebe && \
+    adduser -u 1000 zeebe --system --ingroup zeebe
```
Why do we need this in the builder image?

npepinpe replied:
To preserve permissions when copying over

menski replied:
I don't think you need the user and groups, you can chown to whatever you want as far as I know

npepinpe replied:
I can try that :+1:

menski commented on Dockerfile:
```diff
@@ -16,9 +15,13 @@ RUN mkdir -p ${TMP_DIR} && \
 ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini ${TMP_DIR}/bin/tini
 COPY docker/utils/startup.sh ${TMP_DIR}/bin/startup.sh
 RUN chmod +x -R ${TMP_DIR}/bin/
+RUN chown 1000:1000 -R ${TMP_DIR}/ && chmod -R 0775 ${TMP_DIR}/ ${TMP_DIR}/data
```
I don't think we need to set the owner in the builder image as we set them on copy to the final image

menski commented on Dockerfile:
```diff
@@ -39,21 +42,17 @@ ENV ZB_HOME=/usr/local/zeebe \
+RUN groupadd -g 1000 zeebe && \
+    adduser -u 1000 zeebe --system --ingroup zeebe
+RUN chmod g=u /etc/passwd
+RUN chown -R 1000:1000 ${ZB_HOME}
```
The group has to be `0` for the non-root to work correctly, and I think we should set it after we copied the content, or maybe it's is even not needed, not sure how `copy --chown` works in docker out of my head.

menski commented on Dockerfile:
```diff
@@ -39,21 +42,17 @@ ENV ZB_HOME=/usr/local/zeebe \
-RUN chmod +x ${ZB_HOME}/bin/broker
-RUN chmod +x ${ZB_HOME}/bin/gateway
+USER zeebe
+COPY --from=builder --chown=1000:1000 /tmp/zeebe/bin/startup.sh /usr/local/bin
+COPY --from=builder --chown=1000:1000 /tmp/zeebe ${ZB_HOME}
```
again group has to be `0`, and maybe we can just keep the `startup.sh` in the `${ZB_HOME}/bin` dir instead of copying it around as we also do it `tini`. we can adjust the `ENTRYPOINT` instead.

npepinpe replied:
Can you explain why the group has to be 0? How did you test this? The image runs fine without the 0 group, should I be testing something else?

menski replied:
The group `0` or root group is recommended by the OpenShift documentation, see section https://access.redhat.com/documentation/en-us/openshift_container_platform/4.4/html/images/creating-images#images-create-guide-openshift_create-images paragraph "Support arbitrary user ids" 

> For an image to support running as an arbitrary user, directories and files that may be written to by processes in the image should be owned by the root group and be read/writable by that group. Files to be executed should also have group execute permissions.

> Adding the following to your Dockerfile sets the directory and file permissions to allow users in the root group to access them in the built image:
```
RUN chgrp -R 0 /some/directory && \
    chmod -R g=u /some/directory
```
>Because the container user is always a member of the root group, the container user can read and write these files.

npepinpe commented:
Alright, tested the latest in the cloud with the PSP provided by @coolpalani, and I can create a pod as a restricted user and run Zeebe as user 1000 under fsGroup 1000. One small hiccup is with the logs - by default, the rolling appender is enabled, but as `/usr/local/zeebe` has group 0 and user 1000, if you use any other UID then it won't be able to create logs. There the workaround is to either disable the appender, or set a volume for the logs with custom permissions.

I created a PSP enabled cluster in our Cloud project, you can get the credentials as

```
gcloud container clusters get-credentials np-psp --zone europe-west1-d --project zeebe-io
```

Then save the following as `pod.yml`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: np-test
spec:
  containers:
    - name:  np-test
      image: gcr.io/zeebe-io/zeebe:4788-slim-image
      imagePullPolicy: Always
      volumeMounts:
        - name: data
          mountPath: /usr/local/zeebe/data
      securityContext:
        runAsUser: 2000
        runAsGroup: 2000
  volumes:
    - name: data
      emptyDir: {}
  securityContext:
    runAsUser: 2000
    runAsGroup: 2000
    fsGroup: 2000
```

And create the pod with impersonation:

```
kubectl apply -f pod.yml --as=system:serviceaccount:default:restricted
```

coolpalani commented:
@npepinpe 

If you're looking dynamically to setup uid and gid during deployment then adding user logics has to be completely modify it. 
Because if you just setup random uid in the deployment, while login container bash will show the below error due user doesn’t have an entry in the /etc/passwd file.
~~~
I have no name!
~~~

Also you could try to setup custom permission through Init containers in Kubernetes/OpenShift unfortunately it requires chmod/chgrp capabilities where it doesn't allowed with strict pod security polices.

The bottom line is that production system, users will spin up branded image with all custom configs instead of use open public images.

npepinpe commented:
@coolpalani So it should be enough to simply provide a fixed UID/GID which a user can run as in a non-privileged environment, right? I'm not really sure what would be the use case of running with dynamic users to be honest, so I think this is fine as it is. Thanks for the quick reply!

npepinpe commented:
bors r+
